README.md
Description
===========

Manages ucarp IP failover on Linux.

ucarp is a portable implemenation of the Common Address Redundacny Protocol
which allows multiple hosts on the same local area network to share a set of IP
addresses to provide automatic failover.

Requirements
============

Platforms
-----------
- CentOS 6
- CentOS 7

Attributes
==========

* vid - ID of the virtual server, defaults to 1
* password - ucarp password
* advskew - advertisement skew in seconds, defaults to 1 for master
* advbase - advertisement interval in seconds, defaults to 1
* master - true if this node is the master
* vip - the virtual IP
* netmask - the netmask of the virtual IP
* interface - the interface ucarp manages, defaults to eth0
* bonded_interfaces - the interfaces to bond when bonding is enabled, defaults to eth0, eth1
* bond_mode - the bonding mode when bonding is used

Usage
=====

Add the ucarp recipe to the runlist. Specify site configuration
with node attributes.

Example role for a master:

    name "ucarp_master"
    description "Role for ucarp master"
    run_list([
      "recipe[ucarp]"
    ])

    override_attributes({
      :ucarp => { 
        :password => "sekret",
        :vip => "192.168.122.10",
        :netmask => "255.255.255.0",
        :master => true
      }
    })


CHANGELOG.md
# CHANGELOG for ucarp

This file is used to list changes made in each version of ucarp.

2.0.0 (2018-07-30)
------------------
- Add ucarp::remove

1.0.3 (2017-01-06)
------------------
- Standardization and systemd fix

## 0.0.6:

* Add supported OSs
* Add support for mode 1 failover bonded interfaces

## 0.0.5:

* Flush the IP address off the enslaved interfaces when setting up bonding

## 0.0.4:

* Add support for bonded interfaces

## 0.0.2:

* Interface is now an attribute

## 0.0.1:

* Initial release of ucarp

- - - 
Check the [Markdown Syntax Guide](http://daringfireball.net/projects/markdown/syntax) for help with Markdown.

The [Github Flavored Markdown page](http://github.github.com/github-flavored-markdown/) describes the differences between markdown on github and standard markdown.


spec/spec_helper.rb
require 'chefspec'
require 'chefspec/berkshelf'

ChefSpec::Coverage.start! { add_filter 'ucarp' }

CENTOS_7 = {
  platform: 'centos',
  version: '7.2.1511'
}.freeze

CENTOS_6 = {
  platform: 'centos',
  version: '6.8'
}.freeze

ALL_PLATFORMS = [
  CENTOS_6,
  CENTOS_7
].freeze

RSpec.configure do |config|
  config.log_level = :fatal
end

shared_context 'common_stubs' do
  before do
    stub_data_bag_item('ucarp', 'lb1').and_return(
      id: 'vip-lb1',
      vip_id: '001',
      vip_address: '192.0.2.4',
      bind_interface: 'eth0',
      password: 'secret'
    )
  end
end


spec/databag_spec.rb
require_relative 'spec_helper'

describe 'ucarp::data_bag' do
  ALL_PLATFORMS.each do |plat|
    context "#{plat[:platform]} #{plat[:version]}" do
      cached(:chef_run) do
        ChefSpec::SoloRunner.new(plat) do |node|
          node.normal['ucarp']['data_bag']['cluster'] = 'lb1'
          node.normal['ucarp']['init_type'] = 'systemd'
        end.converge(described_recipe)
      end

      include_context 'common_stubs'

      it do
        expect { chef_run }.to_not raise_error
      end
      it do
        expect(chef_run).to create_file('/etc/ucarp/vip-001.pwd').with(
          content: 'secret',
          owner: 'root',
          group: 'root',
          mode: '0400'
        )
      end
      it do
        expect(chef_run.file('/etc/ucarp/vip-001.pwd')).to notify('service[ucarp]').to(:restart)
      end
      it do
        expect(chef_run).to create_template('/etc/ucarp/vip-001.conf').with(
          source: 'vip.conf.erb',
          owner: 'root',
          group: 'root',
          mode: '0644'
        )
      end
      it do
        expect(chef_run.template('/etc/ucarp/vip-001.conf')).to notify('service[ucarp]').to(:restart)
      end
      it do
        expect(chef_run).to start_service('ucarp').with(
          supports: { status: true, restart: true }
        )
      end
      it do
        expect(chef_run).to enable_service('ucarp').with(
          supports: { status: true, restart: true }
        )
      end
    end
  end
end


spec/default_spec.rb
require_relative 'spec_helper'

describe 'ucarp::default' do
  ALL_PLATFORMS.each do |plat|
    context "#{plat[:platform]} #{plat[:version]}" do
      cached(:chef_run) do
        ChefSpec::SoloRunner.new(plat).converge(described_recipe)
      end
      it do
        expect(chef_run).to install_package('ucarp')
      end
      if plat[:version].to_i == 7
        it do
          expect(chef_run).to create_cookbook_file('/usr/libexec/ucarp/ucarp').with(
            source: 'ucarp',
            owner: 'root',
            group: 'root',
            mode: '0700'
          )
        end
      end
    end
  end
end


recipes/remove.rb
#
# Cookbook:: ucarp
# Recipe:: remove
#
# Copyright:: 2018, Oregon State University
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

ucarp_databag = begin
                  data_bag_item('ucarp', node['ucarp']['data_bag']['cluster'])
                rescue
                  nil
                end

unless ucarp_databag.nil?
  # Prevent notifications from restarting service after it's already been removed
  pwd_r  = find_resource(:file,     "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.pwd")
  conf_r = find_resource(:template, "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.conf")
  run_context.delayed_notification_collection.delete(pwd_r.declared_key)
  run_context.delayed_notification_collection.delete(conf_r.declared_key)

  edit_resource(:file, "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.pwd") do
    action :delete
  end

  edit_resource(:template, "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.conf") do
    action :delete
  end
end

edit_resource(:service, 'ucarp') do
  action [:stop, :disable]
end

edit_resource(:package, 'ucarp') do
  action :remove
end

# Remove patched file
edit_resource(:cookbook_file, '/usr/libexec/ucarp/ucarp') do
  action :delete
end


recipes/default.rb
#
# Cookbook Name:: ucarp
# Recipe:: default
#
# Copyright 2012, Eric Heydrick
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'yum-epel' if platform_family?('rhel')

package 'ucarp'

# Overwrite packaged file to address bug:
# https://bugzilla.redhat.com/show_bug.cgi?id=1568599
cookbook_file '/usr/libexec/ucarp/ucarp' do
  source 'ucarp'
  owner 'root'
  group 'root'
  mode '0700'
  only_if { platform_family?('rhel') && node['platform_version'].to_i == 7 } # CentOS 7
end


recipes/data_bag.rb
#
# Cookbook Name:: ucarp
# Recipe:: data_bag
#
# Copyright 2012, Eric Heydrick
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

include_recipe 'ucarp'

begin
  ucarp_databag = data_bag_item('ucarp', node['ucarp']['data_bag']['cluster'])
rescue
  raise "Could not find data bag for #{node['ucarp']['data_bag']['cluster']}"
end

file "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.pwd" do
  content ucarp_databag['password']
  owner 'root'
  group 'root'
  mode '0400'
  notifies :restart, 'service[ucarp]'
end

template "/etc/ucarp/vip-#{ucarp_databag['vip_id']}.conf" do
  source 'vip.conf.erb'
  owner 'root'
  group 'root'
  mode '0644'
  variables ucarp_databag.to_hash
  notifies :restart, 'service[ucarp]'
end

service 'ucarp' do
  case node['ucarp']['init_type']
  when 'systemd'
    provider Chef::Provider::Service::Systemd
    service_name "ucarp@vip-#{ucarp_databag['vip_id']}.service"
  end
  supports status: true, restart: true
  action [:start, :enable]
end
